File: leetcode_python/917.py
# https://leetcode-cn.com/problems/reverse-only-letters/
# 917. 仅仅反转字母
from curses.ascii import SO
from operator import le
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 917.py
@Time : 2022/02/23 23:16:28
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
'''
class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        s = list(s)
        left, right = 0, len(s) - 1
        while left < right:
            while left < right and not s[left].isalpha():
                left += 1
            while left < right and not s[right].isalpha():
                right -= 1
            if left < right:
                s[left], s[right] = s[right], s[left]
                left += 1
                right -= 1
        return "".join(s)
            
s = Solution()
print(s.reverseOnlyLetters("ab-cd"))

File: leetcode_c_plus_plus/917.cpp
// 917. 仅仅反转字母
// https://leetcode-cn.com/problems/reverse-only-letters/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 917.cpp
@Time : 2022/02/23 23:21:53
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    string reverseOnlyLetters(string s) {
        int left = 0, right = s.size() - 1;
        while(left < right) {
            while(left < right && !isalpha(s[left])) left++;
            while(left < right && !isalpha(s[right])) right--;
            if(left < right) {
                swap(s[left], s[right]);
                left++;
                right--;
            }
        }
        return s;

    }
};